File: packages/ui/src/lib/fuzzy-match.ts
interface FuzzyMatchResult {
    score: number;
    indices: number[];
}

const WORD_CHAR = /[a-z0-9]/i;

/**
 * Case-insensitive greedy subsequence match.
 *
 * Scoring: +1 per matched char, +4 when the match continues the previous
 * one (consecutive), +3 when it lands on a word start (string start or
 * preceded by a non-alphanumeric char). The total is scaled by 100 and the
 * candidate length subtracted so shorter candidates win ties.
 *
 * Greedy matching is not guaranteed to find the highest-scoring alignment;
 * that's an accepted trade-off for the small lists the palette filters.
 *
 * Returns null when `query` is not a subsequence of `text`.
 */
function fuzzyMatch(query: string, text: string): FuzzyMatchResult | null {
    if (query.length === 0) return { score: 0, indices: [] };

    const q = query.toLowerCase();
    const indices: number[] = [];
    let score = 0;
    let searchFrom = 0;

    for (const char of q) {
        // Compare code unit by code unit against the original text so the
        // returned indices stay aligned with it even when toLowerCase()
        // changes a character's code-unit length (e.g. "İ").
        let idx = -1;
        for (let i = searchFrom; i < text.length; i++) {
            if (text[i].toLowerCase() === char) {
                idx = i;
                break;
            }
        }
        if (idx === -1) return null;

        let charScore = 1;
        if (indices.length > 0 && idx === indices[indices.length - 1] + 1) {
            charScore += 4;
        }
        if (idx === 0 || !WORD_CHAR.test(text[idx - 1])) {
            charScore += 3;
        }

        score += charScore;
        indices.push(idx);
        searchFrom = idx + 1;
    }

    return { score: score * 100 - text.length, indices };
}

// FuzzyMatchResult is intentionally not exported — nothing consumes it yet
// (project rule: don't export until necessary).
export { fuzzyMatch };

File: native/Sources/Taskflow/UI/CommandPalette/FuzzyMatch.swift
import Foundation

/// Result of a fuzzy match: the relevance `score` (higher = better) and the matched character
/// `indices` into the ORIGINAL (non-lowercased) candidate string, used to bold matched chars.
/// Ports the return type of `packages/ui/src/lib/fuzzy-match.ts`.
struct FuzzyResult: Equatable {
    let score: Int
    let indices: [Int]
}

/// Case-insensitive greedy subsequence fuzzy matcher.
/// 1:1 port of `fuzzyMatch` in `packages/ui/src/lib/fuzzy-match.ts`.
enum FuzzyMatch {
    /// Returns `nil` when `query` is not a subsequence of `text`. Empty `query` always matches
    /// (score `-text.count`, no indices). Scoring: +1 base per matched char, +4 when the match is
    /// consecutive with the previous match, +3 when the match is at a word start; final score is
    /// `rawScore * 100 - text.count` so shorter candidates win ties.
    nonisolated static func match(_ query: String, _ text: String) -> FuzzyResult? {
        let textChars = Array(text)
        let lowerText = Array(text.lowercased())
        let lowerQuery = Array(query.lowercased())

        var indices: [Int] = []
        var rawScore = 0
        var searchFrom = 0

        for qChar in lowerQuery {
            var found = -1
            var i = searchFrom
            while i < lowerText.count {
                if lowerText[i] == qChar { found = i; break }
                i += 1
            }
            if found == -1 { return nil }

            var charScore = 1
            if let last = indices.last, found == last + 1 { charScore += 4 }      // consecutive run
            if found == 0 || !isWordChar(lowerText[found - 1]) { charScore += 3 }  // word start
            rawScore += charScore
            indices.append(found)
            searchFrom = found + 1
        }

        _ = textChars   // indices index into the original string; lengths match lowerText.
        return FuzzyResult(score: rawScore * 100 - text.count, indices: indices)
    }

    /// `[a-z0-9]` test on an already-lowercased character (matches the TS `WORD_CHAR` regex).
    private nonisolated static func isWordChar(_ c: Character) -> Bool {
        c.isLetter || c.isNumber
    }
}
